game-app/game-core/src/main/java/games/strategy/triplea/delegate/move/validation/MoveValidator.java

```diff
@@ -1450,62 +1452,67 @@ private MoveValidationResult validateParatroops(
-      final Map<Unit, Unit> airTransportsAndParatroops =
-          TransportUtils.mapTransportsToLoad(paratroopsRequiringTransport, airTransports);
-      for (final Unit paratroop : airTransportsAndParatroops.keySet()) {
-        if (paratroop.hasMoved()) {
-          result.addDisallowedUnit("Cannot paratroop units that have already moved", paratroop);
-        }
-        final Unit transport = airTransportsAndParatroops.get(paratroop);
-        if (transport.hasMoved()) {
-          result.addDisallowedUnit("Cannot move then transport paratroops", transport);
+      for (final Unit airTransport : airTransportDependents.keySet()) {
+        if (airTransport.hasMoved()) {
+          result.addDisallowedUnit("Cannot move then transport paratroops", airTransport);
         }
       }
-      final Territory routeEnd = route.getEnd();
-      for (final Unit paratroop : paratroopsRequiringTransport) {
+      final boolean friendlyEnd = Matches.isTerritoryFriendly(player).test(route.getEnd());
+      final boolean canMoveNonCombat =
+          Properties.getParatroopersCanMoveDuringNonCombat(data.getProperties());
+      final boolean isWrongPhase = !isNonCombat && friendlyEnd && canMoveNonCombat;
+      final boolean mustAdvanceToBattle = friendlyEnd && !canMoveNonCombat;
+      for (final Unit paratroop : paratroopersToAirTransports.keySet()) {
         if (paratroop.hasMoved()) {
           result.addDisallowedUnit("Cannot paratroop units that have already moved", paratroop);
         }
-        if (Matches.isTerritoryFriendly(player).test(routeEnd)
-            && !Properties.getParatroopersCanMoveDuringNonCombat(data.getProperties())) {
+        if (mustAdvanceToBattle) {
           result.addDisallowedUnit("Paratroops must advance to battle", paratroop);
         }
-        if (!nonCombat
-            && Matches.isTerritoryFriendly(player).test(routeEnd)
-            && Properties.getParatroopersCanMoveDuringNonCombat(data.getProperties())) {
+        if (isWrongPhase) {
           result.addDisallowedUnit(
               "Paratroops may only airlift during Non-Combat Movement Phase", paratroop);
         }
       }
-      if (!Properties.getParatroopersCanAttackDeepIntoEnemyTerritory(data.getProperties())) {
-        for (final Territory current :
-            CollectionUtils.getMatches(route.getMiddleSteps(), Matches.territoryIsLand())) {
-          if (Matches.isTerritoryEnemy(player).test(current)) {
-            return result.setErrorReturnResult("Must stop paratroops in first enemy territory");
-          }
-        }
+      if (!Properties.getParatroopersCanAttackDeepIntoEnemyTerritory(data.getProperties())
+          && route.getMiddleSteps().stream()
+              .anyMatch(Matches.territoryIsLand().and(Matches.isTerritoryEnemy(player)))) {
+        return result.setErrorReturnResult("Must stop paratroops in first enemy territory");
       }
     }
     return result;
   }
 
+  private Map<Unit, Unit> dependentsToUnitToTransportsMap(
```

Comment:
Is there a typo in the name of this method? Reading it, it sounds like we have 'dependents' -> units' -> 'transports'

Should it be: 'dependentUnitsToTransportMap'?

Reply:
It's actually "dependents" to "unitsToTransportMap", but I agree that it's hard to read.

How about I rename it to convertTransportKeyedMapToLoadedUnitKeyedMap()?